src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs

```diff
@@ -0,0 +1,138 @@
+    {
+        public static readonly AnalyzerManager Default = new AnalyzerManager();
+
+        // This map stores the tasks to compute HostSessionStartAnalysisScope for session wide analyzer actions, i.e. AnalyzerActions registered by analyzer's Initialize method.
+        // These are run only once per every analyzer.
+        private ImmutableDictionary<DiagnosticAnalyzer, Task<HostSessionStartAnalysisScope>> _sessionScopeMap =
+            ImmutableDictionary<DiagnosticAnalyzer, Task<HostSessionStartAnalysisScope>>.Empty;
+
+        // This map stores the tasks to compute HostCompilationStartAnalysisScope for per-compilation analyzer actions, i.e. AnalyzerActions registered by analyzer's CompilationStartActions.
+        // Compilation start actions will get executed once per-each compilation as user might want to return different set of custom actions for each compilation.
+        private readonly ConditionalWeakTable<Compilation, ConcurrentDictionary<DiagnosticAnalyzer, Task<HostCompilationStartAnalysisScope>>> _compilationScopeMap =
+            new ConditionalWeakTable<Compilation, ConcurrentDictionary<DiagnosticAnalyzer, Task<HostCompilationStartAnalysisScope>>>();
+
+        /// <summary>
+        /// Cache descriptors for each diagnostic analyzer. We do this since <see cref="DiagnosticAnalyzer.SupportedDiagnostics"/> is
+        /// a property rather than metadata. We expect it to be cheap and immutable, but we can't force them to be so, we cache them
+        /// and ask only once.
+        /// </summary>
+        private readonly ConditionalWeakTable<DiagnosticAnalyzer, IReadOnlyList<DiagnosticDescriptor>> _descriptorCache =
+            new ConditionalWeakTable<DiagnosticAnalyzer, IReadOnlyList<DiagnosticDescriptor>>();
+
+        private Task<HostCompilationStartAnalysisScope> GetCompilationAnalysisScopeAsync(
+            DiagnosticAnalyzer analyzer,
+            HostSessionStartAnalysisScope sessionScope,
+            Compilation compilation,
+            Action<Diagnostic> addDiagnostic,
+            AnalyzerOptions analyzerOptions,
+            Func<Exception, DiagnosticAnalyzer, bool> continueOnAnalyzerException,
+            CancellationToken cancellationToken)
+        {
+            var compilationActionsMap = _compilationScopeMap.GetOrCreateValue(compilation);
+            return compilationActionsMap.GetOrAdd(analyzer,
+                Task.Run(() =>
+                {
+                    var compilationAnalysisScope = new HostCompilationStartAnalysisScope(sessionScope);
+                    AnalyzerDriverHelper.ExecuteCompilationStartActions(sessionScope.CompilationStartActions, compilationAnalysisScope, compilation,
+                        analyzerOptions, addDiagnostic, continueOnAnalyzerException, cancellationToken);
+                    return compilationAnalysisScope;
+                }, cancellationToken));
+        }
+
+        private Task<HostSessionStartAnalysisScope> GetSessionAnalysisScopeAsync(
```

Comment:
same as the other one. also getTask doesnt require any state that belong to analyzerManager. so why not the method accepts getTask rather than gets all arguments that only used for the getTask?